src/heimdall/worker.cc

```diff
@@ -0,0 +1,903 @@
+
+      // Scale to tile extent and convert to tile coordinates
+      double tile_x = norm_x * projection.tile_extent;
+      double tile_y = norm_y * projection.tile_extent;
+
+      boost::geometry::append(unclipped_line, point_t(tile_x, tile_y));
+    }
+
+    // Clip the line to the tile boundaries using Boost.Geometry
+    // This properly handles lines that cross tile boundaries
+    multi_linestring_t clipped_lines;
+    boost::geometry::intersection(clip_box, unclipped_line, clipped_lines);
+
+    // Skip if no clipped segments (edge is completely outside tile)
+    if (clipped_lines.empty()) {
+      continue;
+    }
+
+    // Process each clipped line segment (there may be multiple if line crosses tile multiple
+    // times)
+    for (const auto& clipped_line : clipped_lines) {
+      // Skip degenerate lines (less than 2 points)
+      if (clipped_line.size() < 2) {
+        continue;
+      }
+
+      // Convert to vtzero points, removing consecutive duplicates
+      std::vector<vtzero::point> tile_coords;
+      tile_coords.reserve(clipped_line.size());
+
+      int32_t last_x = INT32_MIN;
+      int32_t last_y = INT32_MIN;
+
+      for (const auto& pt : clipped_line) {
+        int32_t x = static_cast<int32_t>(std::round(pt.x()));
+        int32_t y = static_cast<int32_t>(std::round(pt.y()));
+
+        // Skip consecutive duplicate points (can happen after rounding)
+        if (x == last_x && y == last_y) {
+          continue;
+        }
+
+        tile_coords.emplace_back(x, y);
+        last_x = x;
+        last_y = y;
+      }
+
+      // Must have at least 2 unique points to create a valid linestring
+      if (tile_coords.size() < 2) {
+        continue;
+      }
+
+      // Check for opposing edge
+      baldr::graph_tile_ptr opp_tile = edge_tile;
+      const DirectedEdge* opp_edge = nullptr;
+      GraphId opp_edge_id = reader_->GetOpposingEdgeId(edge_id, opp_edge, opp_tile);
+
+      // Add feature using the builder
+
+      const volatile baldr::TrafficSpeed* forward_traffic =
+          edge ? &edge_tile->trafficspeed(edge) : nullptr;
+      const volatile baldr::TrafficSpeed* reverse_traffic =
+          opp_edge ? &opp_tile->trafficspeed(opp_edge) : nullptr;
+
+      edges_builder.add_feature(tile_coords, edge_id, edge, opp_edge_id, opp_edge, forward_traffic,
+                                reverse_traffic, edge_info);
+
+      // Collect the end node of this edge for the nodes layer
+      unique_nodes.insert(edge->endnode());
+    }
+  }
+
+  return unique_nodes;
+}
+
+void heimdall_worker_t::build_nodes_layer(vtzero::tile_builder& tile,
+                                          const std::unordered_set<baldr::GraphId>& unique_nodes,
+                                          const TileProjection& projection) {
+  // Create nodes layer builder
+  NodesLayerBuilder nodes_builder(tile);
+
+  // Render unique nodes as points
+  for (const auto& node_id : unique_nodes) {
+    baldr::graph_tile_ptr node_tile;
```

Comment:
could put it outside of this loop and maybe even sort `unique_nodes` to minimize switching the tile